### PR TITLE
Fracture pipeline with multithreaded mesh clipping

### PR DESCRIPTION
**Pull request by LinDadaism**
- add acceleration structure: AABB tree for cells & convex hulls before testing mesh intersection
- add multithreading to mesh clipping
- connect fracture steps into a full pipeline

### CODE REVIEW

File: src/clipper.cpp [hunk 1]
```diff
@@ -1,6 +1,11 @@
 #include "clipper.h"
+#include <iostream>
+#include <chrono>
 
-
+#define MULTITHREAD 
+#define CGAL_HAS_THREADS
```
Comment (LinDadaism) [inline src/clipper.cpp]: remove unused MACROs

File: src/clipper.cpp [hunk 2]
```diff
@@ -177,6 +182,103 @@ bool clipConvexAgainstCell(const MeshConvex& convex, const Cell& cell, spConvex&
...
+            }
+            spConvex result(new MeshConvex{ vertices, faces, convexm, {0, 0, 0}, volume });
+            pattern.getCells()[cellID]->convexes.push_back(result);
+        }
+    }
+}
+#endif
+
+void clipAABB(Compound& compound, Pattern& pattern) {
+    CGAL::Rigid_triangle_mesh_collision_detection<Surface_mesh> tree;
```
Comment (LinDadaism) [inline src/clipper.cpp]: Use `typedef ... ABTree`

File: src/clipper.cpp [hunk 3]
```diff
@@ -177,6 +182,103 @@ bool clipConvexAgainstCell(const MeshConvex& convex, const Cell& cell, spConvex&
...
+#endif
+
+void clipAABB(Compound& compound, Pattern& pattern) {
+    CGAL::Rigid_triangle_mesh_collision_detection<Surface_mesh> tree; 
+    int cell_num = pattern.getCells().size();
+    int convex_num = compound.convexes.size();
+    // put all surface meshes into AABB tree 
+    for (const auto& cell : pattern.getCells()) {
+        // assume all meshes 1 connected pieces
+        //tree.add_mesh(cell->cellMesh, PMP::parameters::apply_per_connected_component(false));
```
Comment (LinDadaism) [inline src/clipper.cpp]: clean duplicate code in comment

File: src/clipper.cpp [hunk 4]
```diff
@@ -177,6 +182,103 @@ bool clipConvexAgainstCell(const MeshConvex& convex, const Cell& cell, spConvex&
...
+        tree.add_mesh(cell->cellMesh, PMP::parameters::apply_per_connected_component(false));
+
+    }
+    for (const auto& convex : compound.convexes) {
+        //tree.add_mesh(convex->convexMesh, PMP::parameters::apply_per_connected_component(false));
```
Comment (LinDadaism) [inline src/clipper.cpp]: clean duplicate code in comment

File: src/clipper.cpp [hunk 5]
```diff
@@ -286,25 +388,58 @@ std::vector<Compound> islandDetection(Compound& old_compound) {
...
+        std::unordered_set<int> curr; 
+        if (cur_convs.find(i) == cur_convs.end()) {
+            curr.insert(i);
+            for (int j = i+1; j < N; ++j) {
+                if (PMP::do_intersect(old_compound.convexes[i]->convexMesh, old_compound.convexes[j]->convexMesh,
+                    PMP::parameters::do_overlap_test_of_bounded_sides(true), PMP::parameters::do_overlap_test_of_bounded_sides(true))) {
+                    curr.insert(j);
+                }
+            }
+            std::vector<std::shared_ptr<MeshConvex>> new_convex;
```
Comment (LinDadaism) [inline src/clipper.cpp]: use `typedef ... spConvex`

File: src/clipper.cpp (hunk 2 above)
Comment (szleo123) [reply src/clipper.cpp]: done

File: src/clipper.cpp (hunk 3 above)
Comment (szleo123) [reply src/clipper.cpp]: done

File: src/clipper.cpp (hunk 4 above)
Comment (szleo123) [reply src/clipper.cpp]: done

File: src/clipper.cpp (hunk 5 above)
Comment (szleo123) [reply src/clipper.cpp]: done

**Pull request merged**